WIP: Added FizzBuzz example

rainbyte opened the pull request:
Port of ruby-qml FizzBuzz example, work in progress.

Related to issue #3.

Thanks in advance!

seanchas116 commented on examples/fizzbuzz/main.c:
```diff
@@ -0,0 +1,189 @@
+        sprintf(fb->result, "%d", i);
+    }
+
+    qmlbind_value params[1] = { qmlbind_value_new_string(strlen(fb->result), fb->result) };
```
Maybe this should be `const qmlbind_value *params[1] = ...`.

seanchas116 commented on examples/fizzbuzz/main.c:
```diff
@@ -0,0 +1,189 @@
+    if (strcmp(name, "input")) {
+        fb->input = (int) qmlbind_value_get_number(value);
+
+        qmlbind_value params[1] = { qmlbind_value_new_number(fb->input) };
```
Here too.

seanchas116 commented on examples/fizzbuzz/main.c:
```diff
@@ -0,0 +1,189 @@
+        const char *str = qmlbind_string_get_chars(qstr);
+        strcpy(fb->result, str);
+
+        qmlbind_value params[1] = { qmlbind_value_new_string(strlen(fb->result), fb->result) };
```
And here.

florianjacob replied:
Also, `qmlbind_value_new_string_cstr` could be used so you don't have to calculate and submit the length by hand.

seanchas116 commented on examples/fizzbuzz/main.c:
```diff
@@ -0,0 +1,189 @@
+{
+    fizzbuzz_t *fb = (fizzbuzz_t *) obj;
+
+    if (strcmp(name, "input")) {
```
Maybe you intended `strcmp(name, "input") == 0`?

seanchas116 commented on examples/fizzbuzz/main.c:
```diff
@@ -0,0 +1,189 @@
+
+        qmlbind_value params[1] = { qmlbind_value_new_number(fb->input) };
+        qmlbind_signal_emitter_emit(fb->signalEmitter, "inputChanged", 1, params);
+    } else if (strcmp(name, "result")) {
```
`strcmp(name, "result") == 0`

seanchas116 commented on examples/fizzbuzz/main.c:
```diff
@@ -0,0 +1,189 @@
+{
+    fizzbuzz_t *fb = (fizzbuzz_t *) obj;
+
+    if (strcmp(name, "input")) {
```
`strcmp(name, "input") == 0`

seanchas116 commented on examples/fizzbuzz/main.c:
```diff
@@ -0,0 +1,189 @@
+
+    if (strcmp(name, "input")) {
+        return qmlbind_value_new_number(fb->input);
+    } else if (strcmp(name, "result")) {
```
`strcmp(name, "result") == 0`

seanchas116 commented on examples/fizzbuzz/main.c:
```diff
@@ -0,0 +1,189 @@
+{
+    fizzbuzz_t *fb = (fizzbuzz_t *) obj;
+
+    if (strcmp(name, "onInputChanged") == 0 && argc == 1) {
```
`onInputChanged` method is not needed. Instead you can call `on_changed` in `fizzbuzz_set_property_cb`.

rainbyte replied:
Even if it could be discarded... How do I connect `onInputChanged` so it reacts to `inputChanged` signal? Calling it directly seems like a workaround to me...

seanchas116 replied:
Actually, automatic connecting to `on<signalName>` block (not method) is just a language feature of QML and Qt itself does not connect  `<signalName>` signal to `on<signalName>` method automatically. (ruby-qml just [emulates Qt' signal system in Ruby](https://github.com/seanchas116/ruby-qml/blob/master/lib/qml/signal.rb)
So we have to call `onInputChanged` by our hand. The simplest way to do that would be calling `on_changed` function.

rainbyte replied:
Hmm, documentation says:
> Such methods can either directly be called from QML or be used as QML signal handlers, as described in Signal and Handler Event System.

So, `Connecting to QML Signals` example (from [here](http://doc.qt.io/qt-5/qtqml-cppintegration-interactqmlfromcpp.html)) cannot be implemented?
I mean using something similar to `connect`, and leaving qml untouched...

seanchas116 replied:
> I mean using something similar to connect, and leaving qml untouched...

Currently no such feature to `connect` signals in libqmlbind... And I think we shouldn't implement it.
Connecting signals from client objects to client objects should be done in client side's signal system to avoid unnecessary dependency on Qt and keep libqmlbind small. For example, ruby-qml [implements own signal system in Ruby](https://github.com/seanchas116/ruby-qml/blob/master/lib/qml/signal.rb). For the languages that has convenient signal-like system such as [Rx](http://reactivex.io/), you can do that even easier.

This is a simple example written in C, in which it's very hard to implement such signal system, so
avoiding connecting signals and calling `on_changed` directly would be a good solution, although it's a workaround.

seanchas116 commented on examples/fizzbuzz/main.c:
```diff
@@ -0,0 +1,189 @@
+
+    qmlbind_value *instance = qmlbind_component_create(component);
+
+    qmlbind_value *width_property =
```
We could delete these lines outputting `width` property.

florianjacob replied:
Yes, in the other example, this just served to show how to access properties from C code. It is not actually neccessary for anything.

seanchas116 commented:
Generally I think this is a great example for libqmlbind because it show the overview of whole system of libqmlbind nicely! :+1:

florianjacob commented on examples/fizzbuzz/main.c:
```diff
@@ -0,0 +1,189 @@
+    } else if (strcmp(name, "result")) {
+        return qmlbind_value_new_string(strlen(fb->result), fb->result);
+    } else {
+        return qmlbind_value_new_null();
```
If the property does not exist, one should return `qmlbind_value_new_undefined()` (this is how http://doc.qt.io/qt-5/qjsvalue.html#property behaves). Great to write examples to find things like this. :smile: Just added a note to qmlbind's documentation about it.

florianjacob commented on examples/fizzbuzz/main.c:
```diff
@@ -0,0 +1,189 @@
+    } else if (strcmp(name, "result")) {
+        qmlbind_string *qstr = qmlbind_value_get_string(value);
+        const char *str = qmlbind_string_get_chars(qstr);
+        strcpy(fb->result, str);
```
`qstr` is owned by the client code, you need to release it after copying it. Which was also missing, seems like I overlooked it because all other `qmlbind_value_get_*` return primitive types.

florianjacob commented on examples/fizzbuzz/main.c:
```diff
@@ -0,0 +1,189 @@
+        fb->input = (int) qmlbind_value_get_number(value);
+
+        qmlbind_value params[1] = { qmlbind_value_new_number(fb->input) };
+        qmlbind_signal_emitter_emit(fb->signalEmitter, "inputChanged", 1, params);
```
`qmlbind_signal_emitter_emit()` takes `const qmlbind_value *const * params` and therefore does not take ownership of the values. After emitting the signal, you need to release the values inside the `params` array.

florianjacob commented on examples/fizzbuzz/main.c:
```diff
@@ -0,0 +1,189 @@
+        strcpy(fb->result, str);
+
+        qmlbind_value params[1] = { qmlbind_value_new_string(strlen(fb->result), fb->result) };
+        qmlbind_signal_emitter_emit(fb->signalEmitter, "resultChanged", 1, params);
```
Again, the string inside the `params` array has to be released after emitting the signal.

florianjacob commented on examples/fizzbuzz/main.c:
```diff
@@ -0,0 +1,189 @@
+
+    char *params1[] = { "input" };
+    qmlbind_metaclass_add_signal(metaclass, "inputChanged", 1, params1);
+    qmlbind_metaclass_add_property(metaclass, "input", "input");
```
second string argument should be `inputChanged`, it has to be the same name that was given to the signal above. `<propertyName>Changed` is just a convention for signal naming, the signal could be named `foobar42` as well.

florianjacob commented on examples/fizzbuzz/main.c:
```diff
@@ -0,0 +1,189 @@
+
+    char *params2[] = { "result" };
+    qmlbind_metaclass_add_signal(metaclass, "resultChanged", 1, params2);
+    qmlbind_metaclass_add_property(metaclass, "result", "result");
```
Again, second string argument should be `resultChanged` to match the signal name.

florianjacob commented on examples/fizzbuzz/main.c:
```diff
@@ -0,0 +1,189 @@
+        on_changed(fb, argv[0]);
+        return qmlbind_value_new_undefined();
+    } else if (strcmp(name, "quit") == 0) {
+        exit(0); // FIXME: should call app.quit()
```
@seanchas116 This is something libqmlbind is actually missing but probably should expose! @rainbyte good catch! :+1: 

Something like `qmlbind_application_quit(int exitcode)`. By using the global `qApp` instance, we wuoldn't even need a `qmlbind_application` parameter, so the client code does not need to hand that reference around.

florianjacob commented on examples/fizzbuzz/main.c:
```diff
@@ -0,0 +1,189 @@
+
+    if (strcmp(name, "onInputChanged") == 0 && argc == 1) {
+        on_changed(fb, argv[0]);
+        return qmlbind_value_new_undefined();
```
You're doing exactly the right thing here. :+1:  but it reminded me to add a note to the documentation that if called functions have no return type, `call_method` callback should return undefined.

florianjacob commented:
Thanks alot for taking your time to write this example by only reading the documentation, this is very helpful to find flaws in it. :+1: 

Also, it's a great reminder why we all write wrappers, because in C, even an example program is something really hard to do without leaking memory etc… :laughing: Do not be deterred by the number of improvement suggestions seanchas and I wrote, you're doing a great job here!

rainbyte commented:
Already updated the code :)
It gives me lots of warnings, but FizzBuzz is somewhat working.

Tell me if there are more fixes I can apply...
Or if I forgot any of the previous ones :P

rainbyte commented:
Rebased to master

florianjacob commented on examples/fizzbuzz/main.c:
```diff
@@ -0,0 +1,189 @@
+
+int main(int argc, char *argv[])
+{
+    qmlbind_application *app = qmlbind_application_new(argc, argv);
```
Just pulled & compiled your branch. I think we should introduce const casts where necessary in C (in C++, they would be done automatically), like I did in the hello world window example: 
```
// explicit const cast required because in plain C99, constness is only converted on the top level.                  
qmlbind_application *app = qmlbind_application_new(argc, (const char* const *)argv);
```

florianjacob commented on examples/fizzbuzz/main.c:
```diff
@@ -0,0 +1,189 @@
+    qmlbind_metaclass *metaclass =
+        qmlbind_metaclass_new(classRef, "FizzBuzz", callbacks);
+
+    char *params1[] = { "input" };
```
`const char *params` and `const char *params2` make the const warnings go away, because then C is able to convert the constness automatically.

florianjacob commented on examples/fizzbuzz/main.c:
```diff
@@ -0,0 +1,189 @@
+
+        on_changed(fb, value);
+
+        const qmlbind_value *params[1] = {
```
In cases where there's a `qmlbind_value` inside `params`, I think it's the best way to declare the values as non-const because the destructor wants a non-const pointer, but cast the array given to `qmlbind_signal_emitter_emit()` to `const qmlbind_value* const *`, similar as proposed for `argv` in the main function.

florianjacob commented:
TODO-List:

 * [ ] Some kind of `unused(parameter)` macro / pragma / whatever to be able to shut up all the warnings about unused variables. Seems like there are some options that are cross-compiler. Maybe `#pragma unused(engine)` or something? See http://stackoverflow.com/questions/17622237/which-is-the-best-way-to-suppress-unused-variable-warning.
 * [ ] Remove the `ENOUGH` variable. Maybe `realloc()` (http://www.cplusplus.com/reference/cstdlib/realloc/) before every string assignment is the way to go?
 * [ ] Expose & use `QApplication::quit()`. I'll prepare a patch to expose it now. :smile:
 * [ ] Steal the screenshot of the fizzbuzz example of ruby-qml. :wink:

florianjacob commented on examples/fizzbuzz/main.c:
```diff
@@ -0,0 +1,189 @@
+    fizzbuzz_t *fb = (fizzbuzz_t *) obj;
+
+    if (strcmp(name, "quit") == 0) {
+        exit(0); // FIXME: should call app.quit()
```
Before quitting, we could print a message like “Quitting… Last FizzBuzz was: %d” to juse the `fb` instance and to justify a little that we use a method on fizzbuzz for quitting and not just using `QtCore.quit()` in QML directly.

rainbyte commented:
Warnings went away... :)

florianjacob commented:
Nice. :+1: `qmlbind_application_exit()` is now exposed, see #37.

florianjacob commented on examples/fizzbuzz/main.c:
```diff
@@ -0,0 +1,206 @@
+
+void on_changed(fizzbuzz_t *fb, const qmlbind_value *input)
+{
+    int i = (int) qmlbind_value_get_number(input);
```
Before getting the number, we should get the string representation and see whether it's an empty string. I think it's very irritating that deleting the current number, which means entering "", is interpreted as 0 and therefore as fizzbuzz, and then displayed as "last fizzbuzz is: " because it uses the value of the text field.

(Yes, this also applies to ruby-qml's version).

florianjacob commented on examples/fizzbuzz/main.qml:
```diff
@@ -0,0 +1,37 @@
+        anchors.margins: 10
+        TextField {
+            placeholderText: "Input"
+            text: "0"
```
When "" is ignored, we can use that as initial text to not have a fizzbuzz from the very beginning.

florianjacob commented on examples/fizzbuzz/main.qml:
```diff
@@ -0,0 +1,37 @@
+        }
+        Text {
+            id: text
+            text: fizzBuzz.result
```
I think it would be nice to give that number or text some context: Like `"Computation result: " + fizzBuzz.result` or similar.

florianjacob commented on examples/fizzbuzz/main.qml:
```diff
@@ -0,0 +1,37 @@
+    FizzBuzz {
+        id: fizzBuzz
+        input: textField.text
+        onInputWasFizzBuzz: lastFizzBuzz.text = "Last FizzBuzz: " + textField.text
```
This should use `fizzBuzz.input` instead of `textField.text` to decouple it from the input representation.

rainbyte commented:
Made more improvements...

FizzBuzz segfaults with qmlbind_application_exit(0) :/
It does not happen when using exit(0).

rainbyte commented:
Hmm, and I think the example should handle also the case when number is too big.

seanchas116 commented on examples/fizzbuzz/main.c:
```diff
@@ -0,0 +1,227 @@
+
+    if (strcmp(name, "quit") == 0) {
+        printf("Quitting… Last result was: %s\n", fb->result);
+        qmlbind_application_exit(0); // FIXME: it segfaults :(
```
`qmlbind_application_exit()` does not exit program immediately, so probably we need to return `qmlbind_value_undefined()` properly here too. (this fixes the segfault in my environment)

florianjacob commented:
That makes sense that we can't interrupt it there, as those handler methods actually are plugged in in the internals of Qt and therefore are not comparable to other code. Let's wait for @seanchas116 though, I just realized that he must have used another option for providing the exit method in `ruby-qml`, because the exposed exit method did not exist before. :laughing:

rainbyte commented:
Well, I've added the `qmlbind_value_undefined` return and segfault does not occur anymore.

rainbyte commented on examples/fizzbuzz/main.c:
```diff
@@ -0,0 +1,227 @@
+        fb->result = realloc(fb->result, sizeof(char) * fb->result_len);
+        if (fb->result == NULL) {
+            printf("Error reallocating memory");
+            qmlbind_application_exit(0);
```
And what about exit here? Could it cause a segfault?

florianjacob replied:
I don't think so, because just doing nothing is a valid option for `on_changed()`.

seanchas116 replied:
`fb->result` is used later, so we should `return` here to avoid segfault.

florianjacob replied:
Of course @seanchas116 is right, sorry, I did not see the rest of the code that's executed in the overview.

seanchas116 commented:
@florianjacob ruby-qml simply calls [`Qt.quit()`](http://doc.qt.io/qt-5/qml-qtqml-qt.html#quit-method) in QML to quit application.

florianjacob commented:
Ah, now I see how it's done. We could do the same way by using `qmlbind_engine_get_globalobject()`, `get_property("qt")`, `call_method("quit")` in series, but it would have the same effects.

rainbyte commented:
Updated again with the missing return...

florianjacob commented:
Looks good to me now. :+1: . Maybe you could squash the bugfix commits to clean up the history, but then I think it's ready! :shipit:

rainbyte commented:
Updated... I've squashed most of the commits.

I've left out commits which changed the qml, because final result is different from ruby-qml.
Maybe screenshot should be updated :/

rainbyte commented:
Updated again... Just renamed a commit

seanchas116 commented:
Looks good :+1:

seanchas116 commented:
Let's get it merged! (screenshot could be updated later)

florianjacob commented:
@rainbyte nicely done. :+1:
